// w-utils@1.0.0/lib/scroll.js
var setInterval2 = require('./other').setInterval2

// 获取滚动条滚上去的高度
function getScrollTop() {
    return (document.documentElement && document.documentElement.scrollTop) || document.body.scrollTop
}

// 获取网页高度
function getOffsetHeight() {
    return (document.documentElement && document.documentElement.offsetHeight) || document.body.offsetHeight
}

// 获取网页可视区域高度
function getClientHeight() {
    return (document.documentElement && document.documentElement.clientHeight) || window.innerHeight
}

// 设置滚动条的位置
function setScroll (val) {
	if (typeof val !== 'number') {
		throw new Error('argument is not number')
		return
	}
    window.scrollTo(0, val)
    return val
}

// 返回网页顶部
function goTop () {
	var sTop = getScrollTop()
	if (sTop > 0) {
		var val = Math.floor(sTop/10*8)
		setScroll(val)
		setInterval2(goTop)
	}
}

// 去网页底部
function goBottom () {
	var sTop = getScrollTop(),
		seeHei = getClientHeight(),
		hei = getOffsetHeight()
	if (sTop + seeHei < hei) {
		var val = sTop + Math.ceil((hei - sTop) / 10 * 1)
		setScroll(val)
		console.log(val)
		setInterval2(goBottom)
	}
}
module.exports = {
	getScrollTop,
	getOffsetHeight,
	getClientHeight,
	setScroll,
	goTop,
	goBottom
}
