// w-components@1.0.0/components/dialog.js
/**
 * @file dialog
 * @author xzh
 * @date 2019-10-17 13:53:04
 */

import { LitElement, html } from 'lit-element'
import { defineComponent } from '../common'

export default defineComponent('dialog')(
  class extends LitElement {
    static get properties() {
      return {
        open: {
          type: Boolean,
          reflect: true,
        }
      }
    }

    constructor() {
      super()

      this.open = false
    }

    show() {
        this.shadowRoot.querySelector('.dialog').show()
    }

    showModal() {
        this.shadowRoot.querySelector('.dialog').showModal()
    }

    close() {
        this.shadowRoot.querySelector('.dialog').close()
    }

    render() {
      return html`
        <dialog
            class="dialog"
            ?open="${this.open}"
        >
            <slot></slot>
        </dialog>
        <style>
            .dialog {
                padding: 0;
                border: none;
            }

            .dialog::backdrop {
                background: rgba(0, 0, 0, 0.6);
            }
        </style>
      `
    }
  }
)
