// w-components@1.0.0/components/left-menu.js
/**
 * @file left-menu
 * @author xzh
 * @date 2019-09-26 09:20:30
 */

import { LitElement, html } from 'lit-element'
import { defineComponent } from '../common'

export default defineComponent('left-menu')(
  class extends LitElement {
    static get properties() {
      return {
        menuWidth: {
          type: String,
        },
        open: {
          type: Boolean,
          reflect: true,
        }
      }
    }

    constructor() {
      super()

      this.menuWidth = '240px'
      this.open = false
    }

    render() {
      return html`
        <div class="overflow">
        <div class="box" style="left: -${this.open ? 0 : this.menuWidth};">
          <div class="menu">
            <slot name="menu"></slot>
          </div>
          <div class="container">
            <slot></slot>
          </div>
        </div>
      </div>
      <style>
        .overflow {
          overflow: hidden;
        }

        .box {
          width: calc(100vw + ${this.menuWidth});
          position: relative;
          display: flex;
          min-height: 100vh;
          transition: left 0.5s;
        }

        .box .menu {
          width: ${this.menuWidth};
          min-height: 100vh;
          overflow-x: auto;
        }

        .box .container {
          width: 100vw;
          min-height: 100vh;
        }
      </style>
      `
    }
  }
)
