// w-components@1.0.0/components/textarea.js
/**
 * @file textarea
 * @author xzh
 * @date 2019-09-25 21:01:28
 */

import { LitElement, html } from 'lit-element'
import { defineComponent } from '../common'

export default defineComponent('textarea')(
  class extends LitElement {
    static get properties() {
      return {
        value: {
          type: String,
          reflect: true,
        },
        placeholder: {
          type: String,
          reflect: true,
        },
        lineHeight: {
          type: Number,
          reflect: true,
        },
        minLine: {
          type: Number,
          reflect: true,
        },
        maxLine: {
          type: Number,
          reflect: true,
        },
        autoFocus: {
          type: Boolean,
          reflect: true,
        },
      }
    }
  
    constructor() {
      super()

      this.value = ''
      this.placeholder = ''
      this.lineHeight = 30
      this.minLine = 1
      this.maxLine = Infinity
      this.autoFocus = false
    }

    firstUpdated() {
      this.textarea = this.shadowRoot.querySelector('textarea')

      this.textarea.value = this.value
      this.textarea.style.minHeight = `${this.minLine * this.lineHeight}px`
      this.textarea.style.maxHeight = `${this.maxLine * this.lineHeight}px`

      this.resetHeight(this.textarea)
    }

    handleInput(e) {
      this.value = e.target.value

      this.resetHeight(e.target)
      this.dispatchEvent(new CustomEvent('input'))
    }

    handleChange(e) {
      this.value = e.target.value

      this.dispatchEvent(new CustomEvent('change'))
    }

    resetHeight(el) {
      el.style.height = 'auto'
      el.style.height = `${el.scrollHeight}px`
    }

    render() {
      return html`
        <textarea
          rows="1"
          placeholder="${this.placeholder}"
          ?autofocus="${this.autoFocus}"
          @input="${this.handleInput}"
          @change="${this.handleChange}"
        ></textarea>
        <style>
          textarea {
            width: 100%;
            padding: 4px 11px;

            font-size: 14px;
            height: auto;
            max-height: 19999px;
            line-height: 1.5;
            color: rgba(0, 0, 0, .65);
            background-color: #fff;
            background-image: none;
            border: 1px solid #d9d9d9;
            border-radius: 4px;
            outline: none;
            resize: none;
            overflow-y: auto;
            transition: all .3s;
            box-sizing: border-box;
          }

          textarea::-webkit-scrollbar {
            width: 0;
            height: 0;
          }

          textarea:hover {
            border-color: #40a9ff;
          }

          textarea:focus {
            border-color: #40a9ff;
          }
        </style>
      `
    }
  }
)
